examples/free_convection.jl

```diff
@@ -1,52 +1,131 @@
+base_dir = parsed_args["output-dir"]
+
+filename_prefix = "free_convection_N" * string(N) * "_Q" * string(heat_flux) *
+                  "_dTdz" * string(dTdz) * "_k" * string(κ) * "_dt" * string(dt) *
+                  "_days" * string(days)
+output_dir = joinpath(base_dir, filename_prefix)
+
+if !isdir(output_dir)
+    println("Creating directory: $output_dir")
+    mkpath(output_dir)
+end
+
+# Adding a second worker/proccessor for asynchronous NetCDF output writing.
 using Distributed
-addprocs(1)  # For asynchronous output writing.
+addprocs(1)
 
-# Apparently I have to do this even when starting julia with "-p 2 --project"
-# otherwise we get "ERROR: LoadError: On worker 2: ArgumentError: Package Oceananigans
-# not found in current path:"
+# We need to activate the Oceananigans environment on all workers if executing
+# this script from the development repository using "julia --project".
 @everywhere using Pkg
 @everywhere Pkg.activate(".");
 
 @everywhere using Oceananigans
 @everywhere using CuArrays
+@everywhere using Printf
 
-# physical constants
-p0 = 1027
-cp = 4181.3
+# Physical constants.
+ρ₀ = 1027    # Density of seawater [kg/m³]
+cₚ = 4181.3  # Specific heat capacity of seawater at constant pressure [J/(kg·K)]
 
-# set simulation parameters
-Nx, Ny, Nz = 128, 128, 128
+# Set more simulation parameters.
+Nx, Ny, Nz = N, N, N
 Lx, Ly, Lz = 100, 100, 100
-Nt, Δt = 1000, 1
-ν, κ = 1e-4, 1e-4
-Tz = 0.01
-bottom_gradient = Tz
-top_flux = 75 / (p0 * cp) #for flux bc this should be positive for cooling
+Δt = dt
+Nt = Int(days*60*60*24/dt)
+ν = κ  # This is also implicitly setting the Prandtl number Pr = 1.
+
+# To impose a flux boundary condition, the top flux imposed should be negative
+# for a heating flux and positive for a cooling flux, thus the minus sign.
+top_flux = -Q / (ρ₀ * cₚ)
 
-# create the model
-model = Model(N=(Nx, Ny, Nz), L=(Lx, Ly, Lz), ν=ν, κ=κ, arch=:GPU, float_type=Float32)
+# Create the model.
+model = Model(N=(Nx, Ny, Nz), L=(Lx, Ly, Lz), ν=ν, κ=κ, arch=GPU(), float_type=Float64)
 
-# set boundary conditions
+# Set boundary conditions.
 model.boundary_conditions.T.z.left = BoundaryCondition(Flux, top_flux)
 model.boundary_conditions.T.z.right = BoundaryCondition(Gradient, bottom_gradient)
-# the default boundary conditions for velocity is free-slip
 
-# set initial condition
-T_prof = 273.15 .+ 20 .+ Tz .* model.grid.zC
-# make initial condition into 3D array
-T_3d = repeat(reshape(T_prof, 1, 1, Nz), Nx, Ny, 1)
-@. T_3d[:, :, 1:round(Int, Nz/2)] += 0.01*randn() #add noise to the top half of the domain for convection
+# Set initial conditions.
+T_prof = 273.15 .+ 20 .+ Tz .* model.grid.zC  # Initial temperature profile.
+T_3d = repeat(reshape(T_prof, 1, 1, Nz), Nx, Ny, 1)  # Convert to a 3D array.
+
+# Add small normally distributed random noise to the top half of the domain to
+# facilitate numerical convection.
+@. T_3d[:, :, 1:round(Int, Nz/2)] += 0.01*randn()
 
 model.tracers.T.data .= CuArray(T_3d)
```

Comment:
It should be sufficient to reshape `T_prof` to shape `(1, 1, Nz)` and write

```julia
model.tracers.T.data .= T_prof
```

The syntax `.=` will launch a kernel that assigns the values of `T_prof` (appropriately broadcast and on the CPU) to `T.data` (on the GPU). It shouldn't be necessary to convert `T_3d` to a `CuArray` beforehand, I don't think.

Reply:
Unfortunately I think assigning the values of an `Array` to a `CuArray` this way leads to a GPU compiler error. I believe it falls under the scope of this issue: https://github.com/JuliaGPU/CuArrays.jl/issues/241
```julia
julia> using CuArrays
julia> x = rand(100, 100, 100);
julia> xg = CuArray(rand(100, 100, 100));
julia> xg .= x;

ERROR: GPU compilation of #23(CuArrays.CuKernelState, CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,3},Tuple{Bool,Bool,Bool},Tuple{Int64,Int64,Int64}}}}) failed
KernelError: passing and using non-bitstype argument

Argument 4 to your kernel function is of type Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,3},Tuple{Bool,Bool,Bool},Tuple{Int64,Int64,Int64}}}}.
That type is not isbits, and such arguments are only allowed when they are unused by the kernel.

Stacktrace:
 [1] check_invocation(::CUDAnative.CompilerJob, ::LLVM.Function) at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/validation.jl:37
 [2] macro expansion at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/driver.jl:174 [inlined]
 [3] macro expansion at /home/alir_mit_edu/.julia/packages/TimerOutputs/7zSea/src/TimerOutput.jl:216 [inlined]
 [4] #codegen#111(::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/driver.jl:171
 [5] #codegen at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/driver.jl:0 [inlined]
 [6] #compile#110(::Bool, ::Bool, ::Bool, ::Function, ::Symbol, ::CUDAnative.CompilerJob) at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/driver.jl:38
 [7] #compile at ./none:0 [inlined]
 [8] #compile#109 at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/driver.jl:22 [inlined]
 [9] compile at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/compiler/driver.jl:22 [inlined] (repeats 2 times)
 [10] macro expansion at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/execution.jl:378 [inlined]
 [11] #cufunction#144(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(CUDAnative.cufunction), ::getfield(GPUArrays, Symbol("##23#24")), ::Type{Tuple{CuArrays.CuKernelState,CUDAnative.CuDeviceArray{Float64,3,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,3},Tuple{Bool,Bool,Bool},Tuple{Int64,Int64,Int64}}}}}}) at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/execution.jl:347
 [12] cufunction(::Function, ::Type) at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/execution.jl:347
 [13] macro expansion at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/execution.jl:174 [inlined]
 [14] macro expansion at ./gcutils.jl:87 [inlined]
 [15] macro expansion at /home/alir_mit_edu/.julia/packages/CUDAnative/B210M/src/execution.jl:171 [inlined]
 [16] _gpu_call(::CuArrays.CuArrayBackend, ::Function, ::CuArray{Float64,3}, ::Tuple{CuArray{Float64,3},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,3},Tuple{Bool,Bool,Bool},Tuple{Int64,Int64,Int64}}}}}, ::Tuple{Tuple{Int64},Tuple{Int64}}) at /home/alir_mit_edu/.julia/packages/CuArrays/qZCAt/src/gpuarray_interface.jl:59
 [17] gpu_call(::Function, ::CuArray{Float64,3}, ::Tuple{CuArray{Float64,3},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{Array{Float64,3},Tuple{Bool,Bool,Bool},Tuple{Int64,Int64,Int64}}}}}, ::Int64) at /home/alir_mit_edu/.julia/packages/GPUArrays/t8tJB/src/abstract_gpu_interface.jl:151
 [18] gpu_call at /home/alir_mit_edu/.julia/packages/GPUArrays/t8tJB/src/abstract_gpu_interface.jl:128 [inlined]
 [19] copyto! at /home/alir_mit_edu/.julia/packages/GPUArrays/t8tJB/src/broadcast.jl:48 [inlined]
 [20] copyto! at ./broadcast.jl:797 [inlined]
 [21] materialize!(::CuArray{Float64,3}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{3},Nothing,typeof(identity),Tuple{Array{Float64,3}}}) at ./broadcast.jl:756
 [22] top-level scope at none:0
```

Reply:
Well, good point! It worked for 1D arrays and I didn't test a 2+D case. 

You can still do

```julia
julia> a = CuArray(rand(4, 4))
4×4 CuArray{Float64,2}:
 0.68224   0.362092  0.720526  0.907365
 0.167808  0.692426  0.371261  0.600049
 0.766858  0.221333  0.671804  0.272827
 0.310367  0.907882  0.497932  0.381382

julia> b = CuArray(reshape(rand(4), (1, 4)))
1×4 CuArray{Float64,2}:
 0.0441818  0.822776  0.739454  0.653864

julia> a .= b
4×4 CuArray{Float64,2}:
 0.0441818  0.822776  0.739454  0.653864
 0.0441818  0.822776  0.739454  0.653864
 0.0441818  0.822776  0.739454  0.653864
 0.0441818  0.822776  0.739454  0.653864
```

Reply:
This motivates developing our field abstraction system a bit so that it's easier for the user to set initial conditions, though.

Reply:
Definitely! This kind of stuff could frustrate a new user.